### ISSUE TITLE
wrong results for wav2vec2 inference

### ISSUE THREAD

**Issue by shiva1393** [pytorch/fairseq#3657]
#### question?
Trained pretrained wav2vec2 base model for english language. Then finetuned for labeled data. After model finetuning used following code to get results 

`python examples/speech_recognition/infer.py fairseq/data --task audio_pretraining --nbest 1 --path checkpoint_best.pt --gen-subset $subset --results-path results_temp --w2l-decoder viterbi --word-score -1 --sil-weight 0 --criterion ctc --labels ltr --max-tokens 4000000 --post-process letter`

Results: 
INFO:__main__:HYPO:T
INFO:__main__:TARGET:ACTOR DEV PATEL SAYS HIS ROLE IN THE FILM SLUMDOG MILLIONAIRE WAS BOTH A
INFO:__main__:___________________
INFO:__main__:HYPO:T
INFO:__main__:TARGET:THE COUNTRYS AFFAIRS AS PART OF A MAJOR SHAKE UP OF THE MILITARY TOP BRASS 

I'm getting only T as output.  I didn't change any code in infer.py

###How i trained model??
1. Initially trained  wav2vec2 base model using following command
fairseq-hydra-train \
    task.data=/path/to/data \
    --config-dir /path/to/fairseq-py/examples/wav2vec/config/pretraining \
    --config-name wav2vec2_base_librispeech

I got checkpoint_pretrained.pt

2. Data preparation for finetunung

train.tsv
/path/english_wav_files
esp_12.wav	49280
esp_13.wav	50000

train.wrd
HELLO SO ANYTHING ELSE NO
ANUSHKA SHARMA CONTINUE MAKING

train.ltr 
H E L L O | S O | A N Y T H I N G | E L S E | N O |
A N U S H K A | S H A R M A | C O N T I N U E | M A K I N G |

3. Now finetuned for labeled data using following command 
fairseq-hydra-train \
    task.data=/path/to/data \
    model.w2v_path=checkpoint_pretrained.pt\
    --config-dir /path/to/fairseq-py/examples/wav2vec/config/finetuning \
    --config-name base_100h

Now i got checkpoint_best.pt then used this model for inference after installing flashlight bindings. 

#how finetuned model log ? How it converging?  This may help to give inputs to me
[INFO] - {"epoch": 1, "update": 0.224, "loss": "1695.09", "ntokens": "7910.48", "nsentences": "117.02", "nll_loss": "25.075", "wps": "1722.8", "ups": "0.22", "wpb": "7910.5", "bsz": "117", "num_updates": "200", "lr": "1.0425e-06", "gnorm": "6068.3", "loss_scale": "0.125", "train_wall": "319", "gb_free": "8.9", "wall": "970"}

[INFO] - {"epoch": 29, "update": 28.218, "loss": "537.408", "ntokens": "7965.03", "nsentences": "115.58", "nll_loss": "7.798", "wps": "4463.3", "ups": "0.56", "wpb": "7965", "bsz": "115.6", "num_updates": "25200", "lr": "3e-05", "gnorm": "322.429", "loss_scale": "0.25", "train_wall": "304", "gb_free": "8.9", "wall": "47018"}

can you please suggest any solution to overcome this issue.
Thank you in advance

**Comment by shiva1393** [comment]
Now i took model from fairseq/wav2vec_small_100h.pt and decoded it. Decoding part working fine. fairseq/wav2vec_small_100h.pt model asking dict.ltr.txt while decoding but finetuned  above instructed model not asking dict.ltr.txt file. I think finetune model i made mistake. Can any one suggest where i'm doing mistake

**Comment by medabalimi** [comment]
I don't see the validation WER and loss scores. How many hours  of data did you train on?

**Comment by shiva1393** [comment]
Hi,
around 150 hours used for finetuning wav2vec2 model and 450 hours used in pretraining.
In above  discussion i didn't mentioned some more files used in finetuning.
├──  train.tsv
├──  valid.tsv
├── dev_other.ltr
├── dev_other.tsv
├── dev_other.wrd
├── dict.ltr.txt
├── train.ltr
├── train.tsv
├── train.wrd

**Comment by medabalimi** [comment]
You probably should train for more epochs. What's the config file / config parameters used for training? Refer to the Wav2Vec 2 paper and the yaml files.

**Comment by shiva1393** [comment]
Thank you @medabalimi,

  Used exact file from fairseq wav2vec2_base_librispeech.yaml(used in wav2vec2 paper), used 2 gpus instead of 64 in config file.

Changes in pretraining : 
In fairseq/models/wav2vec/wav2vec2.py:     conv_feature_layers: str = field(
        default="[(512, 10, 5)] + [(512, 3, 2)] * 4 + [(512,2,2)]+[(512,2,2)]")  removed one stride[(512,2,2)]  to get 160 downsample.

Yes, I will try for more epochs. Thank you i will update if i won't get any output. Please suggest if you feel any changes.

**Comment by shiva1393** [comment]
Finally i understood pretrained model is not properly trained. If i re-train some times weights are exploding.
 @alexeib which hyperparameters i need to change, In issues some peoples suggested to move fp32 instead of fp16. 
  Can you provide some inputs.

**Comment by alexeib** [comment]
are you training on fewer gpus than what we did in the paper? in that case you need to reduce learning rate. when dealing with exploding fp16 the first thing to do is always to try to lower the learning rate

**Comment by shiva1393** [comment]
Used  same hyper-parameters in librispeech_base config file, here lr: [0.0005], 
I'm using 4 GPUs. 
Thank you i will reduce learning rate and check the performance.

**Comment by lixx0105** [comment]
@shiva1393 hi, do you have any update for this problem? thanks~
